### PR TITLE
add mTLS

### LINKED ISSUE #3: use mTLS
### Feature description

use mTLS for secure connection

### PR DESCRIPTION
**Pull request by maha-hajja**
### Description

Fixes #3 

### Quick checks:

- [x] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-grpc-client/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: destination.go [hunk 1]
```diff
@@ -68,20 +74,39 @@ func (d *Destination) Configure(ctx context.Context, cfg map[string]string) erro
...
 		dialOptions = append(dialOptions,
 			bwgrpc.WithBandwidthLimitedContextDialer(bwlimit.Byte(d.config.RateLimit), bwlimit.Byte(d.config.RateLimit), d.dialer))
 	}
-	conn, err := grpc.DialContext(ctx,
+	if !d.config.TLSDisable {
+		// create TLS credentials with mTLS configuration
+		creds := credentials.NewTLS(&tls.Config{
+			Certificates: []tls.Certificate{d.clientCert},
+			RootCAs:      d.caCertPool,
+			MinVersion:   tls.VersionTLS12,
```
Comment (lovromazgon) [inline destination.go:101]: Is there a reason we don't set the minimum version to 1.3?

File: destination.go [hunk 2]
```diff
@@ -68,20 +74,39 @@ func (d *Destination) Configure(ctx context.Context, cfg map[string]string) erro
...
+		}
+	}
 	return nil
 }
 
 func (d *Destination) Open(ctx context.Context) error {
 	dialOptions := []grpc.DialOption{
 		grpc.WithContextDialer(d.dialer),
-		grpc.WithInsecure(), //nolint:staticcheck // todo: will use mTLS with connection
+		grpc.WithTransportCredentials(insecure.NewCredentials()), // todo: will use mTLS with connection
```
Comment (lovromazgon) [inline destination.go:89]: Looks like we don't need this option here anymore.

File: cmd/generate-certs.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply cmd/generate-certs.sh:1]: Maybe even name it `generate-test-certs.sh`, to make it clear it's a utility meant for testing purposes.

File: destination.go (hunk 1 above)
Comment (lovromazgon) [reply destination.go:101]: We are in charge of both sides, client and server, I see no harm in setting this to 1.3.

File: destination.go (hunk 2 above)
Comment (maha-hajja) [reply destination.go:89]: oh I missed this up while solving the conflicts to change the PR base, but should be fine in the server PR

File: cmd/generate-certs.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (lovromazgon) [reply cmd/generate-certs.sh:1]: OR! Put it in the same folder as the certificates, `test/certs`, that might be the best option.

File: README.md [hunk 4]
```diff
@@ -19,10 +19,24 @@ server, then waits for acknowledgments to be received from the server through th
...
-| `url`       | url to gRPC server.                                                    | true     |               |
-| `rateLimit` | the bandwidth limit in bytes/second, use "0" to disable rate limiting. | false    | 0             |
+| name                   | description                                                                          | required                              | default value |
+|------------------------|--------------------------------------------------------------------------------------|---------------------------------------|---------------|
+| `url`                  | url to gRPC server.                                                                  | true                                  |               |
+| `rateLimit`            | the bandwidth limit in bytes/second, use `0` to disable rate limiting.               | false                                 | `0`           |
+| `mtls.disable`         | flag to disable mTLS secure connection, set it to `true` for an insecure connection. | false                                 | `false`       |
```
Comment (lovromazgon) [inline README.md:26]: Flag 🙈

File: config.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+type MTLSConfig struct {
+	// the client certificate path.
+	ClientCertPath string `json:"client.certPath"`
+	// the client private key path.
+	ClientKeyPath string `json:"client.keyPath"`
+	// the root CA certificate path.
+	CACertPath string `json:"CA.certPath"`
+	// flag to disable mTLS secure connection, set it to `true` for an insecure connection.
+	Disable bool `json:"disable" default:"false"`
```
Comment (lovromazgon) [inline config.go:46]: I'd call the option `disabled`, I find verbs for config options a bit strange.

File: config.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// ParseMTLSFiles parses and validates mTLS params values, returns the parsed client certificate, and CA certificate pool,
+// and an error if the parsing fails
+func (c *Config) ParseMTLSFiles() (tls.Certificate, *x509.CertPool, error) {
```
Comment (lovromazgon) [inline config.go:51]: These functions could be moved to `MTLSConfig` now.

File: test/generate-certs.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+# create a directory to store the generated keys and certificates
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+mkdir -p ${SCRIPT_DIR}/../test/certs
```
Comment (lovromazgon) [inline test/generate-certs.sh:5]:
```suggestion
mkdir -p ${SCRIPT_DIR}/certs
```

File: config.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	MTLS MTLSConfig `json:"mtls"`
+}
+
+type MTLSConfig struct {
+	// the client certificate path.
+	ClientCertPath string `json:"client.certPath"`
+	// the client private key path.
+	ClientKeyPath string `json:"client.keyPath"`
+	// the root CA certificate path.
+	CACertPath string `json:"CA.certPath"`
```
Comment (lovromazgon) [inline config.go:44]: I know CA is an abbreviation, I'd still use lowercase `ca` for the config field, we also use `mtls` and `url` for instance.

**Review by lovromazgon — APPROVED** [review]
Only nitpicks left. Looks good 👍

File: config.go (hunk 8 above)
Comment (maha-hajja) [reply config.go:44]: I was thinking about this and convinced myself that `ca` is better and refactored it, then convinced myself that `CA` is better and reverted everything 🫠
BUT the url arguments convinced me again :D

**Pull request merged**